### ISSUE TITLE
"taste" script assumes .bash_profile

### ISSUE THREAD

**Issue by slhck** [alexanderepstein/Bash-Snippets#56]
**Issue Label:**
* [x] Bug
* [ ] New feature
* [ ] Enhancement
* [ ] New component

**Description:**

Using Zsh (through Homebrew), the `taste` script runs successfully after setting the API key in `.zshrc`, but it shows an error:

    /usr/local/bin/taste: line 7: /Users/werner/.bash_profile: No such file or directory

Not sure if a `source ~/.bash_profile` is even needed; if the environment variable is exported properly, there's no need to source a shell startup script. I guess the line can be removed without problems.

PS: Thanks for this collection of scripts!

**OS and OS version:**
* [x] Mac
* [ ] Linux 32 Bit
* [ ] Linux 64 Bit
* [ ] Windows 32 Bit
* [ ] Windows 64 Bit

OS Version: 10.10

**Comment by alexanderepstein** [comment]
Glad you like them!

I don't personally use zsh and I agree that i shouldn't need to source ```~/.bash_profile``` but on my linux distro this is necessary, just retested this by commenting the line and was thrown the no API key error. So it will stay for that reason. 

However before sourcing the file I can check what shell you are in by using the env variable ```$SHELL``` and if its not bash don't source that file. However I think if you are running a bash script no matter what your default shell is this variable will return bash. If that is true then that is not a viable solution and all i can think to do is suppress the error.

**Comment by slhck** [comment]
Have you done `export TASTE_API_KEY=...` and not just `TASTE_API_KEY=...`?

If you export the variable from your shell config file, there's no need to source it when running another script like `taste`.

> but on my linux distro this is necessary

Mind you that only `.bashrc` is read when opening new non-interactive shell sessions. This is why `.bash_profile` isn't read on Linux unless you actually log in. (Read [here](http://www.joshstaiger.org/archives/2005/07/bash_profile_vs.html) for more.)
Technically it's better to place this API key in `.bashrc` under Linux and source `.bash_profile` from `.bashrc . OS X / macOS is different because it reads `.bash_profile` when you open a Terminal.

**Comment by slhck** [comment]
Just to show you that it works without the shell config file (Bash on Linux):

```
werner@OCTOPUS23:~/Downloads$ taste
/usr/local/bin/taste: line 7: /home/werner/.bash_profile: No such file or directory
Error: API key not setup properly
To get an API key visit https://tastedive.com/account/api_access
After getting the API key run the following command: export TASTE_API_KEY="yourAPIKeyGoesHere"
After following all the steps and issues still persist try adding export TASTE_API_KEY manually to your .bash_profile
werner@OCTOPUS23:~/Downloads$ export TASTE_API_KEY=blah
werner@OCTOPUS23:~/Downloads$ taste
/usr/local/bin/taste: line 7: /home/werner/.bash_profile: No such file or directory
Taste
Description: A recommendation engine that provides 3 similar items based on some input topic.
  Taste also has the ability to provide information on the item of interest.
  Supports: shows, books, music, artists, movies, authors, games
Usage: taste [flag] [item]
  -i Get more information on similar items
  -s Get information on the item itself
  -u Update Bash-Snippet Tools
  -h Show the help
  -v Get the tool version
Examples:
  taste -i Kendrick Lamar
  taste Catcher in the Rye
  taste -s Red Hot Chili Peppers
```

**Comment by alexanderepstein** [comment]
I did run it as ``export TASTE_APIKEY=...``` I use elementary OS and basically everything is meant to be protected (I cant even install my own scripts without sudo). I can refain from sourcing the ```bash_profile``` or ```bashrc``` if i export the key and run the taste tool in the same shell. However as soon as i exit the session and open a new one I am hit with the no API key.

Originally I was sourcing ```bashrc``` since I do most of my work on Linux but I switched over after #36 . I guess it would make more sense to check if you are on OSX or Linux and then source the appropriate file. It seems I may have to just supress the output of that source because it clearly doesn't effect the functionality of the tool for you.

**Comment by slhck** [comment]
Hmm. I think it's more of a problem with your setup that requires this hack.

Your script *should* work without sourcing anything. Adding a switch depending on the OS won't help, because you'd have to then cater for every possible combination of OS and shell. What if the user is 
running `fish`, or `zsh` (as in my case)?

Just as a side-note: no other tools (e.g., `git`, `wget`) require you to do this. For example, `export HTTP_PROXY=...` works universally with all tools that read it (like `wget`), as long as the shell you're calling the tool from has that environment variable exported. And this depends on the shell, not the tool.

**Comment by alexanderepstein** [comment]
I only don't think so because I have Elementary OS installed on 3 different laptops all require sudo permissions to do a lot of tasks. I can't seem to find the link but I read a post from an Elementary Developer as to why they did this.

I can't really speak to why it doesn't because i agree with you.

I wouldn't check for all shells only bash and else if bash then source the file suppress the possible errors.

You are also correct that other tools do this in the way that you describe. I am relatively new to bash and I still don't have a full understanding on concepts such as this. Clearly they have a better method for doing this I just don't know how to implement this myself.

**Comment by slhck** [comment]
> Clearly they have a better method for doing this

They don't have a special method for doing it -- they just access `$HTTP_PROXY`. Shells are built in such a way that if you have a parent shell that calls a subshell, it will export its environment variables to that subshell. For example, if you run:

```
export FOO=bar
bash -c 'env' | grep FOO
```

... your parent shell will have `FOO` in its environment, pass that on to `bash`, which is why you can access it from the `bash` subshell. Similarly, when you put such an `export` command into your shell's startup file, it will be available to *any* script you call from that shell.

Thus, make sure that when you run `env` from a freshly launched shell, you can actually see your API key there.

I can't speak for ElementaryOS, but you don't need sudo to edit your own Bash profile (see e.g. [here](http://mstd.eu/index.php/2016/10/06/edit-the-bashrc-file-in-elementary-os/)).

**Comment by alexanderepstein** [comment]
I will have to look into testing this further it seems a lot of the times my OS is causing issues for me that noone else seems to have. Which is fine but like in this case sometimes I implement workarounds for my case but clearly are not necessary for most other *nix machines. I will find a more reliable way to make this work on my distro and other platforms. My quick and dirty solution of suppressing the output will either be out tonight or sometime tomorrow but that won't actually solve the issue.

**Comment by slhck** [comment]
Sure – happy to help, of course, or feel free to post on http://superuser.com/ or https://unix.stackexchange.com/!

A simple

    source ~/.bash_profile 2> /dev/null

should suffice as a workaround.

**Comment by alexanderepstein** [comment]
Implemented workaround c42c66c43902e7c6b2cbb4e4b32e509acf940a79
